doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst

```diff
@@ -5,269 +5,297 @@ Reasoning with inductive types
+      constructor 2.  (* equivalent to "right" *)
 
-   .. tacv:: destruct @natural
+   You can also apply a constructor by name.  (`apply or_intror` will fail):
 
-      :n:`destruct @natural` behaves as :n:`intros until @natural`
-      followed by destruct applied to the last introduced hypothesis.
+   .. coqtop:: reset none
 
-     .. note::
-        For destruction of a number, use syntax :n:`destruct (@natural)` (not
-        very interesting anyway).
+      Goal  forall P1 P2 : Prop, P1 -> P1 \/ P2.
 
-   .. tacv:: destruct @pattern
+   .. coqtop:: all
 
-      The argument of :tacn:`destruct` can also be a pattern of which holes are
-      denoted by “_”. In this case, the tactic checks that all subterms
-      matching the pattern in the conclusion and the hypotheses are compatible
-      and performs case analysis using this subterm.
+      apply or_introl.  (* equivalent to "left" *)
 
-   .. tacv:: destruct {+, @term}
 
-      This is a shortcut for :n:`destruct @term; ...; destruct @term`.
+.. _CaseAnalysisAndInduction:
 
-   .. tacv:: destruct @term as @or_and_intropattern_loc
+Case analysis
+-------------
 
-      This behaves as :n:`destruct @term` but uses the names
-      in :token:`or_and_intropattern_loc` to name the variables introduced in the
-      context. The :token:`or_and_intropattern_loc` must have the
-      form :n:`[p11 ... p1n | ... | pm1 ... pmn ]` with ``m`` being the
-      number of constructors of the type of :token:`term`. Each variable
-      introduced by :tacn:`destruct` in the context of the ``i``-th goal
-      gets its name from the list :n:`pi1 ... pin` in order. If there are not
-      enough names, :tacn:`destruct` invents names for the remaining variables
-      to introduce. More generally, the :n:`pij` can be any introduction
-      pattern (see :tacn:`intros`). This provides a concise notation for
-      chaining destruction of a hypothesis.
+The tactics in this section implement case
+analysis on inductive or co-inductive objects (see :ref:`gallina-inductive-definitions`).
 
-   .. tacv:: destruct @term eqn:@naming_intropattern
-      :name: destruct … eqn:
+Some useful terms:
 
-      This behaves as :n:`destruct @term` but adds an equation
-      between :token:`term` and the value that it takes in each of the
-      possible cases. The name of the equation is specified
-      by :token:`naming_intropattern` (see :tacn:`intros`),
-      in particular ``?`` can be used to let Coq generate a fresh name.
+* `n` is a :gdef:`dependent premise` of `forall n:nat, n + 0 = n` because
+  `n` appears both in the binder of the `forall` and in the quantified statement
+  `n + 0 = n`.
 
-   .. tacv:: destruct @term with @bindings
+* `A` and `B` are :gdef:`non-dependent premises <non-dependent premise>` of `A -> B -> C`
+  because they don't appear in a `forall` binder.
+  (Recall that `A -> B` is notation for the term `forall _ : A, B`; the wildcard
+  `_` can't be referred to in the quantified statement.)
```

Comment:
These are terms of general interest. Therefore, it would make sense to move them to some earlier section of the refman. I'm thinking here: https://coq.github.io/doc/master/refman/language/core/assumptions.html#functions-fun-and-function-types-forall

Reply:
Will do.  I'd like to add "dependent product" and "product", too, if you can provide definitions.

Reply:
I would suggest using the term "product type" instead of just "product" for clarity. Product types are just `forall @ident : @type__1, @type__2` by definition (dependent meaning, again, that `@ident` appears in `@type__2`). They can also be called "function types" given that they correspond exactly to the types of functions (or abstractions as this paragraph call them). Indeed, the type of `fun @ident : @type__1 ==> @term` is `forall @ident : @type__1, @type__2` where `@type__2` is the type of `@term`.

Reply:
Are "abstraction" and "generalization" the same thing?  If so, I think I'd prefer the latter here because we do have a `generalize` tactic.  But both terms are quite common and rarely defined.

Reply:
To my ear, abstraction has a slight connotation of being over a variable, and generalization has the connotation of being over a term.  (That is, abstraction emphasizes the part where you introduce a lambda/pi to close over the abstracted variable, whereas generalization emphasizes the part where you replace a term (which might or might not be just a variable) with a fresh variable.). Or perhaps in some contexts "abstraction" suggests introducing a lambda while "generalization" suggests introducing a forall. But, yes, they are basically the same, and these differences in connotations are very slight.  And I would like Coq to get a reduction strategy called "abstract over", which performs `pattern` without creating the application(s) at the end.

Reply:
I left abstraction alone for now.  `generalize` is not quite what I remembered.